File: Bwl.Framework/src/AutoUI/AutoUIForm.vb
Public Class AutoUIForm
    Inherits FormBase
    Protected _ui As IAutoUI
    Protected _lastUiAlive As DateTime = Now

    Public Sub New()
        MyBase.New
        InitializeComponent()
    End Sub

    Public Sub New(storage As ISettingsStorageForm, logger As ILoggerDispatcher, ui As IAutoUI)
        _storageForm = storage
        _loggerServer = logger
        _ui = ui
        InitializeComponent()
        AddHandler AutoUIDisplay1.AutoFormDescriptorUpdated, AddressOf FormDescriptorUpdated
    End Sub

    Private Sub FormDescriptorUpdated(sender As AutoUIDisplay)
        If Me.IsDisposed Then Return
        Me.Invoke(Sub()
                      Dim desc = AutoUIDisplay1.AutoFormDescriptor
                      Me.Text = desc.Text + " " + desc.ApplicationDescription
                      Me.logWriter.Visible = desc.ShowLogger
                      If logWriter.Visible = False Then
                          Me.AutoUIDisplay1.Height += logWriter.Height
                      End If
                      If desc.FormWidth > 0 Then Width = desc.FormWidth
                      If desc.FormHeight > 0 Then Height = desc.FormHeight
                      logWriter.ExtendedView = desc.LoggerExtended
                  End Sub)
    End Sub

    Public Sub New(appbase As AppBase)
        Me.New(appbase.RootStorage, appbase.RootLogger, appbase.AutoUI)
    End Sub

    Private Sub AutoForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If Me.IsDisposed Then Return
        If Not DesignMode Then
            If _ui IsNot Nothing Then
                AutoUIDisplay1.ConnectedUI = _ui
                _loggerServer.RequestLogsTransmission()
            End If
        End If
    End Sub

    Public Shared Shadows Function Create(appBase As AppBase) As AutoUIForm
        Dim form As New AutoUIForm(appBase)
        Return form
    End Function

    Public Shared Shadows Function Create(storage As ISettingsStorageForm, logger As ILoggerDispatcher, ui As IAutoUI) As AutoUIForm
        Dim form As New AutoUIForm(storage, logger, ui)
        Return form
    End Function

    Private Sub AutoUIForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        If Me.IsDisposed Then Return
        _ui = Nothing
        AutoUIDisplay1.ConnectedUI = Nothing
    End Sub

    Private Sub checkAlive_Tick(sender As Object, e As EventArgs) Handles checkAlive.Tick
        If Me.IsDisposed Then Return
        Dim thr As New Threading.Thread(Sub()
                                            Try
                                                If _ui.CheckAlive() = False Then
                                                    Me.Invoke(Sub()
                                                                  If Text.Contains(" (no connection)") = False Then
                                                                      Text += " (no connection)"
                                                                      ' For Each cnt As Control In Controls
                                                                      ' cnt.Enabled = False
                                                                      ' Next
                                                                  End If
                                                              End Sub)
                                                Else
                                                    Me.Invoke(Sub()
                                                                  If Text.Contains(" (no connection)") = True Then
                                                                      Text = Text.Replace(" (no connection)", "")
                                                                      ' For Each cnt As Control In Controls
                                                                      'cnt.Enabled = True
                                                                      ' Next
                                                                  End If
                                                              End Sub)
                                                End If
                                                _loggerServer.RequestLogsTransmission()
                                            Catch ex As Exception
                                            End Try
                                        End Sub)
        thr.IsBackground = True
        thr.Start()
    End Sub
End Class

File: Framework/Framework/Bwl.Framework.Windows/AutoUI/AutoUIForm.cs
using System;
using System.Runtime.Versioning;
using System.Windows.Forms;

namespace Bwl.Framework.Windows
{

    [SupportedOSPlatform("windows")]
    public partial class AutoUIForm : FormBase, IAutoUIForm
    {

        protected IAutoUI _ui;
        protected DateTime _lastUiAlive = DateTime.Now;

        private string IUIWindow_Text
        {
            get => Text;
            set => Text = value;
        }

        string IUIWindow.Text { get => IUIWindow_Text; set => IUIWindow_Text = value; }

        private event EventHandler _iuiWindowLoad;
        private event EventHandler _iuiWindowFormClosed;

        event EventHandler IUIWindow.Load
        {
            add => _iuiWindowLoad += value;
            remove => _iuiWindowLoad -= value;
        }

        event EventHandler IUIWindow.FormClosed
        {
            add => _iuiWindowFormClosed += value;
            remove => _iuiWindowFormClosed -= value;
        }

        public AutoUIForm() : base()
        {
            InitializeComponent();
        }

        public AutoUIForm(AppBase appbase) : this(appbase.RootStorage, appbase.RootLogger, appbase.AutoUI)
        {
        }

        public AutoUIForm(ISettingsStorageForm storage, ILoggerDispatcher logger, IAutoUI ui)
        {
            _storageForm = storage;
            _loggerServer = logger;
            _ui = ui;
            InitializeComponent();
            AutoUIDisplay1.AutoFormDescriptorUpdated += FormDescriptorUpdated;
        }

        private void FormDescriptorUpdated(object? sender, AutoUIDisplay uiDisplay)
        {
            if (IsDisposed)
                return;
            Invoke(() =>
                {
                    var desc = AutoUIDisplay1.AutoFormDescriptor;
                    Text = desc.Text + " " + desc.ApplicationDescription;
                    logWriter.Visible = desc.ShowLogger;
                    if (logWriter.Visible == false)
                    {
                        AutoUIDisplay1.Height += logWriter.Height;
                    }
                    if (desc.FormWidth > 0)
                        Width = desc.FormWidth;
                    if (desc.FormHeight > 0)
                        Height = desc.FormHeight;
                    logWriter.ExtendedView = desc.LoggerExtended;
                });
        }
        private void AutoForm_Load(object sender, EventArgs e)
        {
            if (IsDisposed)
                return;
            if (!DesignMode)
            {
                if (_ui is not null)
                {
                    AutoUIDisplay1.ConnectedUI = _ui;
                    AutoUIDisplay1.RecreateControls();
                    _loggerServer.RequestLogsTransmission();
                }
            }
        }

        public static new IAutoUIForm Create(AppBase appBase)
        {
            var form = new AutoUIForm(appBase);
            return form;
        }

        public static new IAutoUIForm Create(ISettingsStorageForm storage, ILoggerDispatcher logger, IAutoUI ui)
        {
            var form = new AutoUIForm(storage, logger, ui);
            return form;
        }

        private void AutoUIForm_Load(object sender, EventArgs e)
        {
            _iuiWindowLoad?.Invoke(sender, e);
        }

        private void AutoUIForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (IsDisposed)
                return;
            _ui = null;
            AutoUIDisplay1.ConnectedUI = null;
            _iuiWindowFormClosed?.Invoke(sender, e);
        }

        private void checkAlive_Tick(object sender, EventArgs e)
        {
            if (IsDisposed) return;
            Task.Run(() =>
            {
                try
                {
                    if (_ui.CheckAlive() == false)
                    {
                        Invoke(() =>
                        {
                            if (!Text.Contains(" (no connection)"))
                            {
                                Text += " (no connection)";
                            }
                        });
                    }
                    else
                    {
                        Invoke(() =>
                        {
                            if (Text.Contains(" (no connection)"))
                            {
                                Text = Text.Replace(" (no connection)", "");
                            }
                        });
                    }
                    _loggerServer.RequestLogsTransmission();
                }
                catch (Exception ex) { }
            });
        }

        public void IUIWindow_Dialog(object invokeForm)
        {
            ShowDialog((Form)invokeForm);
        }

        void IUIWindow.ShowDialog(object invokeForm) => IUIWindow_Dialog(invokeForm);

        private void IUIWindow_Show()
        {
            Show();
        }

        void IUIWindow.Show() => IUIWindow_Show();

        private void IUIWindow_Close()
        {
            Close();
        }

        void IUIWindow.Close() => IUIWindow_Close();

        private void IUIWindow_Invoke(Action action)
        {
            Invoke(action);
        }

        void IUIWindow.Invoke(Action action) => IUIWindow_Invoke(action);
    }
}